_source/_posts/2022-03-29-github-actions-graalvm.adoc

```diff
@@ -0,0 +1,490 @@
+    timeout-minutes: 40
+    env:
+      NODE_VERSION: 16.14.0
+      SPRING_OUTPUT_ANSI_ENABLED: DETECT
+      SPRING_JPA_SHOW_SQL: false
+      JHI_DISABLE_WEBPACK_LOGS: true
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v1
+        with:
+          node-version: 16.14.0
+      - uses: actions/setup-java@v2
+        with:
+          distribution: 'temurin'
+          java-version: 11
+      - name: Install node.js packages
+        run: npm install
+      - name: Run backend test
+        run: |
+          chmod +x mvnw
+          npm run ci:backend:test
+      - name: Run frontend test
+        run: npm run ci:frontend:test
+      - name: Package application
+        run: npm run java:jar:prod
+      - name: 'E2E: Package'
+        run: npm run ci:e2e:package
+      - name: 'E2E: Prepare'
+        run: npm run ci:e2e:prepare
+      - name: 'E2E: Run'
+        run: npm run ci:e2e:run
+        env:
+          CYPRESS_ENABLE_RECORD: false
+          CYPRESS_PROJECT_ID: ${{ secrets.CYPRESS_PROJECT_ID }}
+          CYPRESS_RECORD_KEY: ${{ secrets.CYPRESS_RECORD_KEY }}
+      - name: 'E2E: Teardown'
+        run: npm run ci:e2e:teardown
+----
+{% endraw %}
+
+To test this out on your new repository, you'll need to create a branch and pull request (PR) with your changes.
+
+[source,shell]
+----
+git checkout -b actions
+git add .
+git commit -m "Add GitHub Actions"
+git push ci-cd actions
+----
+
+If you watch the tests run from your PR, you'll be pretty pleased until it hits the **E2E: Package** phase. It'll fail with the following error:
+
+----
+Error:  Failed to execute goal org.graalvm.buildtools:native-maven-plugin:0.9.10:build
+  (build-native) on project flickr-2: Execution build-native of goal
+  org.graalvm.buildtools:native-maven-plugin:0.9.10:build failed:
+  GraalVM native-image is missing from your system.
+Error:   Make sure that GRAALVM_HOME environment variable is present.
+----
+
+== The environmental impact of GraalVM builds
+
+This brings a fork in the road. Should you use the https://github.com/graalvm/setup-graalvm/issues/6[setup-graalvm action] to configure GraalVM and your Java SDK?
+
+If you do, every time you create a PR, and commit to it, it will run a native build. These take a 3-4 minutes for me locally, but with GitHub Actions, it takes 30+ minutes! It'll also only build on Linux, so you don't know if there's issues with macOS or Windows.
+
+To me, this seems as bad for the environment as cryptocurrency. It'll also make you wish you bought crypto a couple of years ago if you're using a private repo. You only get 2000 free minutes of GitHub Actions for private repos. Any minutes after that, you get charged for.
+
+_Yes, I know the cryptocurrency topic is controversial. I do like to poke fun at it though. Native builds on every commit and mining bitcoin seem similar to me._
```

Comment:
Then again, simply [surfing the web](https://www.fastcompany.com/90717181/surfing-the-web-isnt-just-bad-for-your-brain-its-terrible-for-the-environment) is terrible for the environment too.